### DISCUSSION TITLE
thread/jmm

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#138] [Announcements]
# thread/jmm

全面理解Java的内存模型（JMM）

https://tobebetterjavaer.com/thread/jmm.html

**Comment by Dawnstar-y** [comment]
看晕了

**Reply by ExcuMe** [reply]
没太看懂这个顺序一致性模型的意义，意思就是只要结果和程序代码顺序执行结果一致，jmm并不关系你做了什么重排序吗？

**Reply by ChuToo968** [reply]
是的，我的理解就是我们写代码的时候以为程序运行时是按照编写的顺序执行的，就是我们想的就是理论的顺序一致性模型。但实际上不是，会有重排序，所以JMM有一系列约束确保了虽然有重排序（优化，提升cpu性能），但不影响结果。也就是JMM不保证按我们想的顺序执行（顺序一致性模型执行），但保证执行结果一致。同步程序和未同步程序我的理解就是多线程，并发的时候，我们自己有没有利用锁机制，volatile，synchronized等去自行处理线程同步，没处理也就是未同步程序，那执行的时候JMM也没保证。就是文中说得：
JMM 的一系列约束 的 平衡点：对编译器和处理器来说，只要不改变程序的执行结果（单线程程序和正确同步了的多线程程序），编译器和处理器怎么优化都行。

**Comment by ExcuMe** [comment]
没太看懂这个顺序一致性模型的意义，意思就是只要结果和程序代码顺序执行结果一致，jmm并不关系你做了什么重排序吗？

**Reply by DavidYin66** [reply]
感觉这写的就没想让你看懂
